### PR TITLE
Support SM4-NI optimization

### LINKED ISSUE #250: 基于AESNI进行SM4加速

### PR DESCRIPTION
**Pull request by InfoHunter**
SM4-CTR/GCM can be optimized by utilizing AES-NI. It would have a boost on performance at +130% for normal size data blocks. For instance, on 256 byte data blocks, we have:

    118MB/s (Tongsuo-8.3) vs 285MB/s (Tongsuo-8.3 w/SM4NI)

The optimized version is nearly 2.5 times faster of the former one.

This feature requires compilers that have 'x86intrin.h' to get built.

通常情况下8.3-stable分支并不接受新的特性，但是此优化较为重要，因此计划合并进8.3分支。

此功能基于 Markku-Juhani O. Saarinen 的理论：https://github.com/mjosaarinen/sm4ni

This PR resolves #250 

##### Checklist

- [x] 在 https://yuque.com/tsdoc 增加或更新了必要的文档：https://www.yuque.com/tsdoc/ts/eksg4u092ixd349k
- [x] 增加或更新了必要的测试用例
- [x] 对于重要修改，更新了CHANGES文件
- [ ] 当前修改存在对已有API参数或返回值的改变
- [ ] 当前修改存在对旧版本功能的兼容性改变（如网络协议或密码算法）

### CODE REVIEW

**Comment by InfoHunter** [comment]
Ping the original author @william-zk for a review

**Comment by zzl360** [comment]
could this be assembly language like https://github.com/torvalds/linux/blob/master/arch/x86/crypto/sm4-aesni-avx-asm_64.S and https://github.com/torvalds/linux/blob/master/arch/x86/crypto/sm4-aesni-avx2-asm_64.S since the openssl and Tongsuo use plasm？
another question:as @uudiin mentioned in https://github.com/torvalds/linux/commit/a7ee22ee1445c7fdb00ab80116bb9710ca86a860    ,cbc decrypt could also be optimized  by this code.
and, maybe we should use this code in https://github.com/Tongsuo-Project/Tongsuo/blob/master/providers/implementations/ciphers/cipher_sm4_gcm_hw.c to be used in sm4_gcm.

File: crypto/evp/e_sm4.c [hunk 1]
```diff
@@ -502,7 +505,11 @@ static int sm4_gcm_init(EVP_CIPHER_CTX *ctx, const unsigned char *key,
     if (key) {
         SM4_set_key(key,&gctx->ks);
         CRYPTO_gcm128_init(&gctx->gcm, &gctx->ks, (block128_f) sm4_128_block_encrypt);
+# ifdef USE_SM4_NI
+        gctx->ctr = (ctr128_f) sm4_ctr128_encrypt_blocks;
```
Comment (zzl360) [inline crypto/evp/e_sm4.c:511]: should use cpu feature detection

**Comment by InfoHunter** [comment]
> could this be assembly language like https://github.com/torvalds/linux/blob/master/arch/x86/crypto/sm4-aesni-avx-asm_64.S and https://github.com/torvalds/linux/blob/master/arch/x86/crypto/sm4-aesni-avx2-asm_64.S since the openssl and Tongsuo use plasm？

It could be, but pure assembly language is hard to maintain. It's also not easy to find a long-term maintainer if the original author is no long with the project (OpenSSL also suffered from this in the past). Using compiler intrinsics has less problems but will lose independency of build, which I think is acceptable for Tongsuo's scenarios.

> another question:as @uudiin mentioned in https://github.com/torvalds/linux/commit/a7ee22ee1445c7fdb00ab80116bb9710ca86a860 ,cbc decrypt could also be optimized by this code.
and, maybe we should use this code in https://github.com/Tongsuo-Project/Tongsuo/blob/master/providers/implementations/ciphers/cipher_sm4_gcm_hw.c to be used in sm4_gcm.

That file is in master branch and I will work on that later. For CBC decryption, it can be done but with a low priority. TLS 1.3 has dropped CBC mode and we also intend to make SM4 GCM more popular with high perf in both encryption and decryption.

**Comment by InfoHunter** [comment]
Dead CPU architectures like HP-PA, Alpha and PowerPC are also removed in cross-compile test.

File: include/crypto/sm4.h [hunk 2]
```diff
@@ -33,6 +33,33 @@ int SM4_set_key(const uint8_t *key, SM4_KEY *ks);
...
+ *
+ * GCC > 4.5.0
+ * Recent clang, including Apple clang
+ * Intel icc
+ *
+ * You can't use MSVC to build this feature since it has no 'x86intrin.h'.
+ */
+# ifndef OPENSSL_NO_SM4_NI
+#  if defined (__has_include) && (__has_include(<x86intrin.h>))
+#  include <x86intrin.h>
```
Comment (dongbeiouba) [inline include/crypto/sm4.h]: include需要增加缩进

**Review by dongbeiouba — CHANGES_REQUESTED** [review]

**Review by dongbeiouba — APPROVED** [review]
LGTM

File: crypto/evp/e_sm4.c (hunk 1 above)
Comment (zzl360) [reply crypto/evp/e_sm4.c:511]: should add line
if(AESNI_CAPABLE &&  (OPENSSL_ia32cap_P[1]&(1<<(9))))//aesni and ssse3.but may be should use ssse4_1 which is (OPENSSL_ia32cap_P[1]&(1<<(19))

Comment (InfoHunter) [reply crypto/evp/e_sm4.c:511]: It seems this should be added

**Pull request merged**